fgpyo/io/io.py

```diff
@@ -0,0 +1,178 @@
+
+    Args:
+        paths: list of one or more Paths to be investigated
+
+    Example:
+    _directory_exists(path = Path("/example/directory/"))
+    """
+    assert path.exists(), f"Path does not exist: {path}"
+    assert path.is_dir(), f"Path exists but is not a directory: {path}"
+
+
+def paths_are_writeable(path: Path, parent_must_exist: bool = True) -> None:
+    """Checks that path is writable and returns True, else raises an Exception
+    Args:
+         path: Path to be investigated
+    Example:
+    _writeable(path = Path("example.txt"))
+    """
+    assert path.exists(), f"Cannot write file because path is non-existent: {path}"
+    assert path.is_file(), f"Cannot write file because path is a directory: {path}"
+    assert os.access(path, os.W_OK), f"File exists but is not writebale: {path}"
+
+    if parent_must_exist:
+        parent: str = f"{path.parent.absolute()}"
+        assert Path(
+            parent
+        ).exists(), f"Cannot write file because parent diretory does not exist: {path}"
+        assert Path(
+            parent
+        ).is_dir(), f"Cannot write file because parent exists and is not a directory: {path}"
+        assert os.access(
+            parent, os.W_OK
+        ), f"Cannot write file because parent directory is not writeable: {path}"
+
+
+def to_reader(
+    path: Path, mode: str = "rt"
+) -> Union[gzip.GzipFile, io.TextIOWrapper, TextIO, IO[Any]]:
+    """Opens a Path for reading with a specifeied mode or default 'rt'.
+
+    Args:
+        path: Path to be read from
+        mode: Mode to open reader, default 'rt'
+
+    Example:
+    >>> reader = io.to_reader(path = Path("reader.txt"))
+    >>> reader.readlines()
+    >>> reader.close()
+    """
+    if path.suffix in COMPRESSED_FILE_EXTENSIONS:
+        return gzip.open(path, mode=mode)
+    else:
+        return path.open(mode=mode)
+
+
+def to_writer(path: Path, mode: str = "wt") -> Union[gzip.GzipFile, IO[Any], io.TextIOWrapper]:
+    # TODO find compatible return type
+    """Opens a Path for reading and based on extension uses open() or gzip.open()
+
+    Args:
+        path: Path to write to
+        mode: Mode to open writer, default 'wt'
+
+    Example:
+    >>> writer = io.to_writer(path = Path("writer.txt"))
+    >>> writer.write(f"{something}\n")
+    >>> writer.close()
+    """
+    if path.suffix in COMPRESSED_FILE_EXTENSIONS:
+        return gzip.open(path, mode=mode)
+    else:
+        return path.open(mode=mode)
+
+
+def read_lines(path: Path, strip: bool = True) -> Any:
+    """Takes a path and reads it into a list of strings, removing line terminators
+    along the way.
+    # TODO Make line strip optional
+
+    Args:
+        path: Path to read from
+        strip: Boolean to strip lines or not
+
+    Example:
+    >>> read_back: List[str] = io.read_lines(path)
+    """
+    reader = to_reader(path=path)
+    list_of_lines = reader.readlines()
+    reader.close()
```

Comment:
1. why not use a `with`?
2. why not return a generator so you don't have to read in all the lines into memory?  The scala version returns an iterator for this reason
3. use f-strings rather than `str` (it's faster)
4. use `.rstrip("\r\n")` for stripping trailing line terminators (supports windows)

```python
with to_reader(path=path) as reader:
  if strip:
    for line in reader:
       yield f"{line}".strip()
  else:
    for line in reader:
       yield f"{line}".rstrip("\r\n")
```